### ISSUE TITLE
Metric Depth Extraction

### ISSUE THREAD

**Issue by BenSaunders27** [DhruvJawalkar/Depth-Map-Prediction-from-a-Single-Image-using-a-Multi-Scale-Deep-Network#1]
Hi Dhruv,

I have been able to re-implement your code provided here to generate the depth maps of an image, using the pre-trained model weights provided.

However, I am unsure of how to calculate the exact metric depth for this image? As in the actual distances to objects in the image.

I would expect this to be encoded into the 'pred' variable outputted from the Model, but the pixel values can often be negative, which is logically incorrect.

Is it possible to use this model to extract the pixel-wise metric depths from an image? If so, how is this achieved?

Thanks,
Ben

**Comment by DhruvJawalkar** [comment]
Hi Ben,

So currently,
- The model scales down raw depth values to *0 - 255* for training (In the
dataset class)
- Final 'pred' tensor has normalized values, hence -ve.
- I scale them back to *0 - 255 *by adding min() and scaling

I've actually updated the notebook now to do this and show the plot.

But, if you want to keep the raw depth values as is and *not scale* to *0 -
255, *you could :
- Try if just adding the min() value to the tensor, gives accurate range
results
- Just comment the part for scaling in the dataset class and try retraining
a bit. (you'd need a global depth max value)


Do have and look and sure do let me know if you have any questions.


Thanks,
Dhruv

**Comment by DhruvJawalkar** [comment]
<img width="879" alt="updated-pred-plot" src="https://user-images.githubusercontent.com/5371976/51150294-83258000-181a-11e9-835e-2ca3f042ba14.png">
<img width="801" alt="dataset" src="https://user-images.githubusercontent.com/5371976/51150295-83258000-181a-11e9-8905-2e4a3edef513.png">

**Comment by BenSaunders27** [comment]
Dhruv,

Thanks a lot for your responses, it is helpful. I shall try and re-train the model using the raw metric depths as you suggest.

Just to confirm, when you say "Just comment the part for scaling in the dataset class", this refers to the following lines in 'dataset.py' (30-32)?

depth = (depth/depth.max())*255
depth = depth.astype(np.uint8)

Thanks

**Comment by DhruvJawalkar** [comment]
Hi Ben,

I think this is hard task asking a NN to do, though the NYU dataset does
provide the raw depth values and that you could train it, I feel you might
not get great results.

Predicting relative depth is hard enough for it to do, predicting actual
values from just an image is something even a human expert can't do.

So I feel '//depth = (depth/depth.max())*255'  changing something here
won't do much.

**Comment by aimanian** [comment]
Hi Dhruv,

I am trying to calculate the root mean square error between the predicted (transformed) depth and and the real (transformed) depth from running your code. I used the following code right after predictions to calculate root mean square error (rmse) between predicted and real depths, but the rmse seems to be very high. I was wondering if you can let me know if I am wrong in rmse calculation. Actually, the rmse reported in the paper is much lower, and I think I am doing a mistake.

Many thanks,

**-First I transformed the predictions to 0-255 as you described as,**

pred_numpy_arr = pred.detach().squeeze(dim=1).squeeze(dim=0).cpu().numpy()
pred_numpy_arr = pred_numpy_arr - pred_numpy_arr.min()       # -0.660375
pred_numpy_arr = (pred_numpy_arr/pred_numpy_arr.max())*255   # max = 2.1548061, brings scale to 0-255  
pred_numpy_arr = pred_numpy_arr.astype('uint8')

**- Second I used to following code for rmse calculation for batch size of 8v as:**
from sklearn.metrics import r2_score 
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from math import sqrt
#from math import mean

r2=[]
mae=[]
a=[]
b=0
for i in range(bs):
    a.append(((depth.numpy()[i]-pred_numpy_arr[i])**2).mean())
    r2.append(r2_score(depth.numpy()[i],pred_numpy_arr[i]))
    mae.append(mean_absolute_error(depth.numpy()[i],pred_numpy_arr[i]))
    b=b+a[0]
rms=sqrt(b/8)
rms

**I get rmse=75**
